### PR TITLE
#273 do samples and templates testing on site

### PR DESCRIPTION
**Pull request by CAMOBAP**
- #273
 - https://github.com/metanorma/ci/pull/101 - related PR

## Intro

The approach with testing samples and templates by repository_dispatch doesn't look convenient. Because the track is loosed which metanorma-cli trigger build on mn-sample-x or mn-template-x

## Proposal

Do testing on the gem side (as implemented in this PR). Example is https://github.com/metanorma/metanorma-cli/actions/runs/4044552061

We will test all flavors, and several template doctypes they all will be listed in static JSON files

Benefits of this approach:
 - These tests are implemented in the separate workflow so we still can reuse workflow from `metanorma/ci`
 - Many of our deliverables share the same matrices, makes sense to share it across `metanorma-cli`, `homebrew-metanorma`, `metanorma-chocolatey`, `metanorma-docker`

## Open questions

 - [ ] Should we also add a different OS to the test matrix?
 - [ ] Should we also add a different ruby versions to the test matrix?
 - [ ] Should we test all possible combinations for templates or current subset is enough?

## Next proposal

I think it makes sense to implement the same approach for metanorma-x gems to test relevant samples and templates.
@opoudjis @ronaldtse what do you think?

### CODE REVIEW

**Comment by ronaldtse** [comment]
> I think it makes sense to implement the same approach for metanorma-x gems to test relevant samples and templates.

I fully agree. Let's take this approach.

File: .github/workflows/samples-matrix.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
+{
```
Comment (ronaldtse) [inline .github/workflows/samples-matrix.json:1]: Some whitespace formatting issues in this JSON, can we clean it up?

**Comment by ronaldtse** [comment]
> Should we also add a different OS to the test matrix? 

In the metanorma-* repos, we should test against all OSes.

In the final deliverable here (the final test for release), we should also test against all OSes.

> Should we also add a different ruby versions to the test matrix?

In the metanorma-* repos, we should test against all Ruby versions.

In the final deliverable here (the final test for release), we should also test against all Ruby version.

> Should we test all possible combinations for templates or current subset is enough?

In the metanorma-* repos, we should test against all templates.

In the final deliverable here (the final test for release), we probably only need to test a subset.

**Comment by opoudjis** [comment]
Let me do a release this way before it is set in stone, because if this ends up meaning that my releases will involve ANY dead time waiting between gem releases, then I will reject it violently. ("Violently" meaning "I will not do it, and will IGNORE the test results.") I will not be sitting around waiting for jobs to finish for two hours, to trigger the next job. I initiate releases 7 pm as is.

The sequence I release gems in is by dependency, as shown in the metanorma-release readme:

- isodoc
- metanorma
- metanorma-standoc
- metanorma-iso
- metanorma-general
- gems dependent only on standoc: ietf itu ieee nist ogc un
- gems dependent on metanorma-iso: iec bsi
- gems dependent on metanorma-general and metanorma-iso: bipm
- gems dependent on metanorma-general: csa cc iho m3waag ribose
- metanorma-cli

The dependencies that would slow down because of interspersed samples testing are:

* metanorma-iso > metanorma-iec
* all samples > metanorma-cli docker final run

The metanorma-iso > metanorma-iec gap is probably long enough that this won't be an issue, given that all of general ietf itu ieee nist ogc un need to be released between them.

The prospect on the other hand that I will need to wait until the entire backlog of samples tests runs, before I can even push metanorma-cli prerelease, is noxious.

@ronaldtse says:

> In the metanorma-* repos, we should test against all OSes.
> In the final deliverable here (the final test for release), we should also test against all OSes.
> In the metanorma-* repos, we should test against all Ruby versions.
> In the final deliverable here (the final test for release), we should also test against all Ruby version.
> In the metanorma-* repos, we should test against all templates.

It is very easy for @ronaldtse to say that, because he's not the one who will be waiting for all of these to run for hours. Do not assume that all of these will survive my initial test execution.

And you will NOT include Alpine in testing. This takes too long already, I'm not going to accept cycles of testing being wasted on a nice-to-have OS.

**Review by opoudjis — APPROVED** [review]
I'll try it out, but I'm going to be very intolerant of inefficiencies. Releases are already at the edge of what is feasible for me to do in one night, and these will likely push way past that.

**Comment by CAMOBAP** [comment]
@ronaldtse I have addressed all feedback from your https://github.com/metanorma/metanorma-cli/pull/295#issuecomment-1415285229

I really support the idea that we should do good testing before release. But also I understand point of @opoudjis I think we should find some compromise here.

If we are going to test:
- ruby: 2.7 .. 3.2
- os: macOS, Windows, Ubuntu
- flavors, templates (reduced)

It will take hours to pass, and the release will probably take more than one day

My proposal:
- significantly reduce flavors and keep only the ruby version which we use in tamtebako/docker
- @opoudjis feel free to delegate the release of some gems to me or if CI takes look time don't hesitate to ask me to watch out. Because we have well-documented release steps in 

Questions:
- because templates usually more simple can we do smoke/ci tests with templates only?

**Comment by ronaldtse** [comment]
> My proposal:
> 
> * significantly reduce flavors and keep only the ruby version which we use in tamtebako/docker

I agree to only keep the "current" Ruby version which is used in Tebako and Docker.

> * top 5 flavors: bipm, ogc, itu, iso, ieee-private. which takes ~20 min to run, maybe we can skipe some of them or restrict number of documents in metanorma.yml to compile

Instead of limiting flavors, I would rather limit the number of documents per flavor being compiled (i.e. we select a few representative samples).

> * @opoudjis feel free to delegate the release of some gems to me or if CI takes look time don't hesitate to ask me to watch out. Because we have well-documented release steps in

Can we actually automate the release process?

**Comment by CAMOBAP** [comment]
> I agree to only keep the "current" Ruby version which is used in Tebako and Docker.

👍 , I will update PR

> Instead of limiting flavors, I would rather limit the number of documents per flavor being compiled (i.e. we select a few representative samples).

Maybe you can list suggest representative docs from those flavors?

> Can we actually automate the release process?

### Intro

Well, what we have so far:
1. almost on all gems we have "a button to release"
   Screenshot 2023-02-07 at 3 08 17 PM
1. we have a formal description for our gem repositories and dependencies between them https://github.com/metanorma/metanorma-release/blob/main/lapidist.yaml
1.  some time ago I thought that https://github.com/metanorma/lapidist can simplify this. But this is just CLI tool it's not about automation

Let me try to design a better version of automation releases right here:

### Requirements:
 - One button release (fire and forget). similar way as we can release for a single repository/gem (N1 from the list above)
 - Independent releases are still possible and will not trigger a cascade release of other gems
 - Clear track point of failure, and notification about this
 - Recoverability. If the release failed after some point we should be able to fix the issue and continue a failed release
 - Don't do useless releases (if no commit was pushed)

### Possible implementation
 - Now I think about the possibility to implement such an approach by some "hypervisor" GH Actions workflow where each job is responsible to trigger release for each gem, some pseudo-code of how such workflow can look like:
   ```yaml
    ...
    jobs:
       isodoc:
         steps:
           # trigger workflow_dispatch
           # wait for the release finish isodoc's workflow should trigger somehow metanorma-release's workflow somehow
           # make sure that gem is published to rubygems.org
       metanorma:
           steps:
           # trigger workflow_dispatch, TODO update dependencies version in gemspec
           # wait for the release finish. TODO metanorma's workflow should trigger somehow metanorma-release's workflow somehow
           # make sure that gem is published to rubygems.org
       metanorma-standoc:
         needs: [isodoc, metanorma]
           # trigger workflow_dispatch, TODO update dependencies version in gemspec
           # wait for the release finish. TODO isodoc's workflow should trigger somehow metanorma-release's workflow somehow
           # make sure that gem is published to rubygems.org
       ...
       metanorma-cli:
         needs: [metanorma-iso. metanorma-standoc, metanorma ...]
   ```

There are still a few open questions about the approach above that looks working. I have double-checked the limits https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits and they looks good to us. Especially if we are going one by one for some 

This approach sounds really promising to me:
 - it will not reduce release time
 - it will distract the person who is responsible for the release much less 
 - it's maintainable enough.
 - we can start with PoC and trigger release for a few gems: `isodoc` and `metanorma` for instance
 - we still will need to release some gems "manually" like `metnorma-utils`

This is really a raw concept, but I really would like to hear your feedback @ronaldtse @opoudjis
If you see some potential in it I will create a dedicated ticket and start research/PoC

**Comment by ronaldtse** [comment]
@CAMOBAP you know I'm all for this concept. We've discussed about this for *years*, and maybe it is now the right time to make it work!

**Comment by opoudjis** [comment]
In principle agreement :)

The documents have been picked somewhat randomly for the samples repos. I will draw up a list, but I am not going to spend hours doing so. I am going to aim for:

* At least one document from every subfolder in the repo
* And at least a quarter of all documents in the repo

We can always (I hope!) add documents later, so long as the listing of documents to include in initial testing is on some file somewhere I can edit.

This is what I think we should use. It's not a meticulous selection, but it's what you're getting.

### mn-samples-bipm (8 of 31):

sources/JCTLM/jctlm-ec-07.adoc
sources/brochure-its90/brochure-its90-fr.adoc
sources/guide-its90/guide-its90-01.adoc
sources/guides/guide-its90-diff.adoc
sources/working-party-notes/wpn-203.adoc
sources/cipm-mra/d-02/document.adoc
sources/monographie/bipm-1/document.adoc
sources/rapports/2015-01/document.adoc

### mn-samples-bsi (5 of 16):

sources/bs-en-iso-13485-2012/document.adoc
sources/flex-0/document.adoc
sources/pas_2060/document.adoc
sources/bsi-iso-20400-44000-kit/bs-11000-2/document.adoc
sources/bsi-iso-44000-kit/pd-iso-tr-44000/document.adoc

### mn-samples-cc (1 of 1):

sources/cc-18011.adoc

### mn-samples-csa (1 of 1):

sources/csa-dso/document.adoc

### mn-samples-iec (1 of 3):

sources/iec-rice-en.adoc

### mn-samples-iec-private (1 of 3):

sources.mnconvert/iec62830-8{ed1.0}en/document.adoc

### mn-samples-ieee (2 of 2):

sources/p987.6-amd/document.adoc	
sources/p987.6/document.adoc

### mn-samples-ieee-private (1 of 4):

sources/1127/document.adoc

### mn-samples-ietf (1 of 4):

sources/antioch-v2/document.adoc

### mn-samples-iho (3 of 9):

sources/s5-guidelines/document.adoc
sources/b12/document.adoc
sources/s102/document.adoc

### mn-samples-iso (8 of 28):

sources/amendment/rice-amd-en.awi.adoc
sources/guide/document.adoc
sources/international-standard/rice-en.prf.adoc
sources/international-workshop-agreement/document.adoc
sources/publicly-available-specification/document.adoc
sources/technical-report/document.adoc
sources/technical-specification/document.adoc
sources/directives/part2/document.adoc

### mn-samples-itu (6 of 23):

sources/G.191.adoc
sources/T-Editing-Guidelines-201602.adoc
sources/T-REC-H.782-201811-I.MSW-E.adoc
sources/T-RES-T.1-2016-MSW-E.adoc
sources/T-TUT-CABLETV-2020.adoc
sources/Z.100-201811-AnnF1.adoc

### mn-samples-jcgm (1 of 4):

sources/jcgm/100-2008-en/document.adoc

### mn-samples-m3aawg (3 of 3):

sources/best-practice/document.adoc	
sources/policy/document.adoc		
sources/report/document.adoc

### mn-samples-nist (4 of 12):

sources/NIST.CSWP.01262018.adoc
sources/NIST.SP.800-116r1.covers.draft-wip.adoc
sources/NIST.SP.800-116r1.covers.withdrawn-final.adoc
sources/NIST.SP.800-90B.adoc

### mn-samples-ogc (5 of 19):

sources/00-027/document.adoc
sources/05-020r27/document.adoc
sources/08-062r7/document.adoc
sources/10-091r3/document.adoc
sources/style-sample/document.adoc

### mn-samples-ribose (1 of 2):

sources/nereon-syntax/rsd-nereon-configuration-syntax.adoc

### mn-samples-un (3 of 10):

sources/ECE_AGAT_2020_INF1.adoc		
sources/ECE_TRADE_437E_Rec42.adoc	
sources/ECE_TRADE_C_CEFACT_2018_1E.adoc

**Comment by CAMOBAP** [comment]
@opoudjis thanks a lot, I will prepare "smoke manifests"

**Comment by CAMOBAP** [comment]
@abunashir @opoudjis @ronaldtse all suggestions applied. Could you please review one more time. Please just approve/comment I will need to do some changes before merge (I will merge myself)

**Comment by CAMOBAP** [comment]
@opoudjis @abunashir @ronaldtse I still would like to get your feedback on this PR

**Comment by ronaldtse** [comment]
I agree.

**Comment by opoudjis** [comment]
I don't want to do this now, because I have just started release, and this will be another bad one. But approving.

**Review by opoudjis — APPROVED** [review]

**Comment by CAMOBAP** [comment]
@opoudjis I will merge after current release

**Pull request merged**